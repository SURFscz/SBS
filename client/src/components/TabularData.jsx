import React, {Fragment} from "react";
import "./TabularData.scss";
import I18n from "../locale/I18n";
import DOMPurify from "dompurify";
import {isEmpty} from "../utils/Utils";
import {dateColumns, requiredColumns} from "../utils/CSVParser";
import {Tooltip} from "@surfnet/sds";

export default function TabularData({
                                        headers = [],
                                        data = [],
                                        errors = [],
                                        showRequiredInfo = true,
                                        isResultView = false,
                                        allEmails=[]
                                    }) {

    const errorTranslation = (row, error) => {
        const hasErrorTranslation = I18n.translations[I18n.locale].bulkUpload.errors[error.code];
        return I18n.t(`bulkUpload.errors.${hasErrorTranslation ? error.code : "Unknown"}`, {
            fields: requiredColumns.filter(field => isEmpty(row[field])).join(", "),
            invitee: error.invitee,
            shortName: error.shortName,
            message: error.message
        });
    }

    const errorClassName = error => {
        return (error.code || "").indexOf("Warning") > -1 ? "warning" : "error";
    }

    const displayHeader = header => {
        if (requiredColumns.includes(header)) {
            return header + "<sup class='required'>*</sup>";
        }
        return header;
    }

    const displayValue = (header, value) => {
        if (Array.isArray(value)) {
            if (header === "invitees" && isResultView) {
                return (
                    <div className="invitees-container">
                    {value.map((val, index) => <span key={index}
                                                      className={allEmails.includes(val) ? "": "strike-through"}>{val}</span>)}
                </div> );

            }
            return value.join(", ");
        }
        if (dateColumns.includes(header) && !isEmpty(value)) {
            const isoString = new Date(parseInt(value, 10) * 1000).toISOString();
            return isoString.substring(0, 10);
        }
        return value || "";
    }

    const trClassName = error => {
        const resultPart = isResultView ? "results" : "";
        const errorPart = error ? (error.code || "").indexOf("Warning") > -1 ? "warning-row" : "error-row" : "";
        return `${errorPart} ${resultPart}`
    }

    return (
        <div className="tabular-data">
            <table>
                <thead>
                <tr>
                    <th className="index"></th>
                    {headers.map((header, index) =>
                        <th key={index} className={header}>
                            <span dangerouslySetInnerHTML={{__html: DOMPurify.sanitize(displayHeader(header))}}/>
                            {showRequiredInfo && <Tooltip tip={I18n.t(`bulkUpload.tooltips.${header}`)}/>}
                        </th>
                    )}
                </tr>
                </thead>
                <tbody>
                {data.map((row, index) => {
                    const error = errors.find(error => error.row === index);
                    return <Fragment key={index}>
                        <tr className={trClassName(error)}>
                            <td>{index + 1}</td>
                            {headers.map((header, innerIndex) =>
                                <td key={innerIndex}>{displayValue(header, row[header])}</td>
                            )}
                        </tr>
                        {error &&
                            <tr>
                                <td className={errorClassName(error)}
                                    colSpan={headers.length + 1}>
                                    {errorTranslation(row, error)}
                                </td>
                            </tr>}
                    </Fragment>
                })}
                </tbody>
            </table>
            {showRequiredInfo &&
                <p className="info"><sup className='required'>*</sup><em>{I18n.t("bulkUpload.requiredInfo")}</em></p>}
        </div>
    );

}

