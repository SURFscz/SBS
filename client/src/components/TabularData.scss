@import "../stylesheets/vars.scss";

.tabular-data {
    margin-bottom: 25px;

    table {
        width: 100%;
        border-collapse: collapse;
        font-family: Arial, sans-serif;
        font-size: 14px;
    }

    th, td {
        border: 1px solid #d4d4d4;
        padding: 8px;
        text-align: left;
    }

    th {
        background-color: #f3f3f3;
        font-weight: bold;

        &.short_names {
            width: 12%;
        }

        &.intended_role {
            width: 12%;
        }

        &.invitees {
            width: 10%;
        }

        &.groups {
            width: 10%;
        }

        &.invitation_expiry_date {
            width: 17%;
        }

        &.membership_expiry_date {
            width: 17%;
        }

        &.message {
            width: 10%;
        }

        &.sender_name {
            width: 15%;
        }

    }

    tr:nth-child(even) {
        background-color: #f9f9f9;
    }

    tr.results:not(.error-row) td {
        color: var(--sds--color--green--400);
    }

    tr.error-row td {
        color: var(--sds--color--red--400);
    }

    td.error {
        color: white;
        background-color: var(--sds--color--red--400);

        &:hover {
            background-color: var(--sds--color--red--500);
        }
    }

    input {
        border: none;
        width: 100%;
        padding: 4px;
        font-size: 14px;
        background: transparent;
    }

    input:focus {
        outline: 2px solid #4a90e2;
        background-color: #fff;
    }

    p.info {
        margin-top: 8px;
        font-size: 14px;
    }
}
