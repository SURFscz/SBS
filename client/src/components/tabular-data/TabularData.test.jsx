import React from "react";
import {render} from "@testing-library/react";
import TabularData from "./TabularData";

describe("TabularData", () => {
    it("matches snapshot with minimal props", () => {
        const {asFragment} = render(
            <TabularData
                data={[]}
                columns={[]}
                onRowClick={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with data and columns", () => {
        const data = [
            { id: 1, name: "Item 1", value: "Value 1" },
            { id: 2, name: "Item 2", value: "Value 2" }
        ];
        const columns = [
            { field: "id", headerName: "ID" },
            { field: "name", headerName: "Name" },
            { field: "value", headerName: "Value" }
        ];
        const {asFragment} = render(
            <TabularData
                data={data}
                columns={columns}
                onRowClick={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
