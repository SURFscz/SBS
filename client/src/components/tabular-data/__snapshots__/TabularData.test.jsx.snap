// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabularData matches snapshot with data and columns 1`] = `
<DocumentFragment>
  <div
    class="tabular-data"
  >
    <table>
      <thead>
        <tr>
          <th
            class="index"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          class=" "
        >
          <td>
            1
          </td>
        </tr>
        <tr
          class=" "
        >
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>
    <p
      class="info"
    >
      <sup
        class="required"
      >
        *
      </sup>
      <em>
        indicates a required value
      </em>
    </p>
  </div>
</DocumentFragment>
`;

exports[`TabularData matches snapshot with minimal props 1`] = `
<DocumentFragment>
  <div
    class="tabular-data"
  >
    <table>
      <thead>
        <tr>
          <th
            class="index"
          />
        </tr>
      </thead>
      <tbody />
    </table>
    <p
      class="info"
    >
      <sup
        class="required"
      >
        *
      </sup>
      <em>
        indicates a required value
      </em>
    </p>
  </div>
</DocumentFragment>
`;
